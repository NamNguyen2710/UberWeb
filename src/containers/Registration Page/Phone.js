import { Form, Field, withFormik } from 'formik';
import * as yup from 'yup';

const Phone = (props) => {
  return (
    <div className="signupForm">
      <h2>Get moving with Uber</h2>
      <Form className="col">
        <div className="input-row">
          <Field className="input-box" as="select" name="postalCode">
            { props.postalCodeList.map(value => <option value={value} key={value}>{value}</option>) }
          </Field>
          <Field 
            className="input-box"
            type="tel" 
            name="phoneNumber" 
            placeholder="Enter your mobile number" 
          />
        </div>
        <label>
          <Field
            type="checkbox" 
            name="agreeTerm"
          />
          Agree Terms and Conditions
        </label>
        <button className="round-btn" type="submit">â†’</button>
      </Form>
    </div>
  );
}

const PhoneForm = props => {
  const SignUpValidation = yup.object().shape({
    postalCode: yup
      .string()
      .required()
      .oneOf([...props.postalCodeList, "Choose your country postal code."]),
    phoneNumber: yup
      .number()
      .min(8)
      .max(10)
      .required(),
    agreeTerm: yup
      .boolean()
      .oneOf([true], "The terms and conditions must be accepted."),
  });

  const PhoneFormWithFormik = withFormik({
    mapPropsToValues: () => ({
      postalCode: props.postalCodeList[1],
      phoneNumber: "",
      agreeTerm: false,
    }),
    handleSubmit: (values, bag) => bag.props.handleSubmit(values, bag.setSubmitting),
    // validationSchema: SignUpValidation,
  })(Phone);
  return <PhoneFormWithFormik {...props} />
}

export default PhoneForm;
