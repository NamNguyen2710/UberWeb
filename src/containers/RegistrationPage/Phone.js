import { Form, Field, ErrorMessage, withFormik } from 'formik';
import * as yup from 'yup';

const Phone = (props) => {
  return (
    <div className="signupForm">
      <Form className="col">
        <h2>Get moving with Uber</h2>
        <div className="input-row">
          <Field className="input-box" as="select" name="postalCode">
            { props.postalCodeList.map(value => <option value={value} key={value}>{value}</option>) }
          </Field>
          <Field 
            className="input-box"
            type="tel" 
            name="phoneNumber" 
            placeholder="Enter your mobile number" 
          />
        </div>
        <ErrorMessage name="phoneNumber">
          { msg => <div className="error-msg">{msg}</div> }
        </ErrorMessage>
        <label>
          <Field
            type="checkbox" 
            name="agreeTerm"
          />
          Agree Terms and Conditions
        </label>
        <ErrorMessage name="agreeTerm">
          { msg => <div className="error-msg">{msg}</div> }
        </ErrorMessage>
        <button className="round-btn" type="submit">â†’</button>
      </Form>
    </div>
  );
}

const PhoneForm = props => {
  const SignUpValidation = yup.object().shape({
    postalCode: yup
      .string()
      .required()
      .oneOf([...props.postalCodeList, "Choose your country postal code"]),
    phoneNumber: yup
      .number("Please enter valid phone number")
      .min(10000000, "Please enter valid phone number")
      .max(999999999999, "Please enter valid phone number")
      .required("Please enter your phone number"),
    agreeTerm: yup
      .boolean()
      .oneOf([true], "The terms and conditions must be accepted."),
  });

  const PhoneFormWithFormik = withFormik({
    mapPropsToValues: () => ({
      postalCode: props.postalCodeList[1],
      phoneNumber: "",
      agreeTerm: false,
    }),
    handleSubmit: (values, bag) => bag.props.handleSubmit(values, bag.setSubmitting),
    validationSchema: SignUpValidation,
  })(Phone);
  return <PhoneFormWithFormik {...props} />
}

export default PhoneForm;
