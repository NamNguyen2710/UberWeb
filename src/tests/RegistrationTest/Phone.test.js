import React from "react";
import { mount } from "enzyme";
import { waitFor } from "@testing-library/react";
import renderer from "react-test-renderer";
import Phone from "../../containers/RegistrationPage/Phone";

describe("sign up phone form", () => {
  it("render correctly", () => {
    const handleSubmit = jest.fn();
    const postalCodeList = [];
    const snapshot = renderer
      .create(
        <Phone handleSubmit={handleSubmit} postalCodeList={postalCodeList} />
      )
      .toJSON();

    expect(snapshot).toMatchSnapshot();
  });

  it("submit to get error", async () => {
    const handleSubmit = jest.fn();
    const postalCodeList = ["+84", "+1"];
    const wrapper = mount(
      <Phone handleSubmit={handleSubmit} postalCodeList={postalCodeList} />
    );

    const phoneField = wrapper.find('input[name="phoneNumber"]');
    const submitBtn = wrapper.find("button");

    phoneField.simulate("focus");
    phoneField.simulate("change", {
      target: { value: "123 456", name: "phoneNumber" },
    });
    phoneField.simulate("blur");
    await waitFor(() => {
      expect(wrapper.find('input[name="phoneNumber"]').props().value).toBe(
        "123 456"
      );
    });

    submitBtn.simulate("submit");
    await waitFor(() => {
      expect(
        wrapper.find("ErrorMessageImpl[name='phoneNumber'] div").exists()
      ).toBeTruthy();
      expect(
        wrapper.find("ErrorMessageImpl[name='agreeTerm'] div").exists()
      ).toBeTruthy();
      expect(handleSubmit).not.toHaveBeenCalled();
    });
  });

  it("correct input phone then submit", async () => {
    const handleSubmit = jest.fn();
    const postalCodeList = ["+84", "+1"];
    const wrapper = mount(
      <Phone handleSubmit={handleSubmit} postalCodeList={postalCodeList} />
    );

    const postalCodeField = wrapper.find('select[name="postalCode"]');

    const phoneField = wrapper.find('input[name="phoneNumber"]');
    const agreeTermField = wrapper.find('input[name="agreeTerm"]');
    const submitBtn = wrapper.find("button");

    postalCodeField.simulate("focus");
    postalCodeField.simulate("change", {
      target: { value: "+1", name: "postalCode" },
    });
    postalCodeField.simulate("blur");
    await waitFor(() => {
      expect(wrapper.find('select[name="postalCode"]').props().value).toBe(
        "+1"
      );
    });

    phoneField.simulate("focus");
    phoneField.simulate("change", {
      target: { value: "123 456 7890", name: "phoneNumber" },
    });
    phoneField.simulate("blur");
    await waitFor(() => {
      expect(wrapper.find('input[name="phoneNumber"]').props().value).toBe(
        "123 456 7890"
      );
    });

    agreeTermField.simulate("change", {
      target: { value: true, name: "agreeTerm" },
    });
    await waitFor(() => {
      expect(wrapper.find('input[name="agreeTerm"]').props().checked).toBe(
        true
      );
    });

    expect(wrapper.find(".error-msg").exists()).toBeFalsy();

    submitBtn.simulate("submit");
    await waitFor(() => {
      expect(handleSubmit).toHaveBeenCalled();
    });
  });
});
