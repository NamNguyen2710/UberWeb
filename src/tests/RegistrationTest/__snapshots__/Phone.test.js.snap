// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sign up phone form render correctly 1`] = `
<div
  className="signupForm"
>
  <form
    action="#"
    className="col"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <h2>
      Get moving with Uber
    </h2>
    <div
      className="input-row"
    >
      <select
        className="input-box"
        name="postalCode"
        onBlur={[Function]}
        onChange={[Function]}
      />
      <input
        className="input-box"
        name="phoneNumber"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter your mobile number"
        type="tel"
        value=""
      />
    </div>
    <label>
      <input
        checked={false}
        name="agreeTerm"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
        value={false}
      />
      Agree Terms and Conditions
    </label>
    <button
      className="round-btn"
      type="submit"
    >
      â†’
    </button>
  </form>
</div>
`;
